### PR TITLE
Newton proximal solver and Newton primal-dual solver for epigraph projection

### PR DESCRIPTION
**Pull request by xflash96**

### CODE REVIEW

File: src/epsilon/prox/logistic.cc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  static Eigen::VectorXd hessf(const Eigen::VectorXd &x) {
+    int n = x.rows();
+    Eigen::VectorXd h(n);
+    for(int i=0; i<n; i++)
+      h(i) = std::exp(x(i)) / pow(1+std::exp(x(i)), 2);
+    return h;
+  }
+};
+
+// lam*||x||_2
```
Comment (mwytock) [inline src/epsilon/prox/logistic.cc]: update

File: src/epsilon/prox/logistic.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  Eigen::VectorXd Apply(const Eigen::VectorXd& v) override {
+    return ProxByNewton(v, lambda_);
+  }
+
+private:
+  double lambda_;
+};
+REGISTER_PROX_OPERATOR(LogisticProx);
+
+// I(||x||_2 <= t)
```
Comment (mwytock) [inline src/epsilon/prox/logistic.cc]: update

File: src/epsilon/prox/logistic.cc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    int n = x.rows();
+    Eigen::VectorXd h(n);
+    for(int i=0; i<n; i++)
+      h(i) = std::exp(x(i)) / pow(1+std::exp(x(i)), 2);
+    return h;
+  }
+};
+
+// lam*||x||_2
+class LogisticProx final : public NewtonProx{
```
Comment (mwytock) [inline src/epsilon/prox/logistic.cc]: small nit: space before {

File: src/epsilon/prox/negentr.cc [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    int n = x.rows();
+    Eigen::VectorXd h(n);
+    for(int i=0; i<n; i++)
+      h(i) = 1/x(i);
+    return h;
+  }
+};
+
+// \sum_i xi log xi
+class NegEntrProx final : public NewtonProx{
```
Comment (mwytock) [inline src/epsilon/prox/negentr.cc]: can you call this NegativeEntropyProx and the fine negative_entropy.cc (same for epigraph)

File: src/epsilon/prox/neglog.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    int n = x.rows();
+    Eigen::VectorXd h(n);
+    for(int i=0; i<n; i++)
+      h(i) = 1 / (x(i)*x(i));
+    return h;
+  }
+};
+
+// \sum_i log(xi)
+class NegLogProx final : public NewtonProx{
```
Comment (mwytock) [inline src/epsilon/prox/neglog.cc]: similar comment: can you call this NegativeLogProx(), and negative_log.cc

File: src/epsilon/prox/logistic.cc (hunk 3 above)
Comment (xflash96) [inline src/epsilon/prox/logistic.cc]: got it.

File: src/epsilon/prox/negentr.cc (hunk 4 above)
Comment (xflash96) [inline src/epsilon/prox/negentr.cc]: Sure.

File: src/epsilon/prox/neglog.cc [hunk 6]
```diff
@@ -0,0 +1,75 @@
+#include "epsilon/affine/affine.h"
+#include "epsilon/expression/expression_util.h"
+#include "epsilon/prox/prox.h"
+#include "epsilon/prox/newton.h"
+#include "epsilon/vector/dynamic_matrix.h"
+#include "epsilon/vector/vector_util.h"
+#include <cmath>
+
+class NegLog {
```
Comment (mwytock) [inline src/epsilon/prox/neglog.cc]: yeah you could probably make NegativeLog, NegativeEntropy and Logistic all inherit from some base class to remove some of the repetition (or you could use templates) but not necessary to do now. Basic any smooth elementwise function can follow this pattern

Comment (xflash96) [inline src/epsilon/prox/neglog.cc]: Just to want avoid multiple inheritance. Func defines (f, g, h). But FuncProx and FuncEpi all need the definition of Func and the basic ProxOperator class. Thus, now I use static class method to give f, g, and h, which avoid multiple inheritance by embedding. Suggestion is welcomed.

Comment (mwytock) [inline src/epsilon/prox/neglog.cc]: Yeah definitely should avoid multiple inheritance. I think if you just did
a minor refactoring to define a class

class SmoothFunction {
  virtual double f(const Eigen::VectorXd &x) = 0;
  virtual Eigen::VectorXd gradf(const Eigen::VectorXd &x) = 0;
  virtual Eigen::VectorXd hessf(const Eigen::VectorXd &x) = 0;
}

and then have NewtonProx/NewtonEpigraph use that (e.g. pass a
SmoothFunction in the constructor)

class Logistic final : public SmoothFunction {

}

Haven't thought the all the details yet. It also seems like you could also
make NewtonProx/NewtonEpigraph templated, e.g.

class LogisticProx : NewtonProx<Logistic> {
 ...
}

which might cut down the lines of code even more ...

**Pull request merged**